### DISCUSSION TITLE
support for virtual machine

### DISCUSSION THREAD

**Discussion by brainchild0** [swizzin/swizzin#649] [Q&A + Help] [answered]
The high level of core administrative manipulation currently required to deploy Swizzin may not be suitable for all host systems. Since Docker containers are not supported, a virtual machine is the remaining option for users wishing to avoid substantial changes on the host. 

Are there any potential pitfalls that may be relevant to a new user attempting deployment through VM? Not accounting for storage of torrent payloads, is any advice available on how small a footprint is reasonable, for storage and memory, in the VM configuration (assuming no more than five users and throughput no more than 5 Mb/s)?

**Comment by flying-sausages** [comment, marked as answer]
I cannot think of any pitfalls for VMs. I use VMs to test whether swizzin runs right myself.

**Comment by brainchild0** [comment]
Do you have any recommendations for memory or storage, to keep the footprint as low as reasonable? (Payload data of course would sit directly on the host filesystem.)

**Reply by flying-sausages** [reply]
This is where swizzin kinda shines. Swizzin itself uses no resources when idle when nothing is installed, as it is just a set of scripts. So your performance really needs to only respond to whatever you specifically want to run. If that's just ZNC, you probably need next to nothing. If you want Plex + sonarr + radarr + qbittorrent, I'd look at 4 cores at the very least, 4GB+ of RAM, and 10GB of storage (given torrents would be outside of the VM). Your mileage may vary based on the speed of your RAM and CPU.

**Reply by brainchild0** [reply]
I just want a torrent engine. Maybe sonarr and raddar at some point, but not soon.

**Reply by flying-sausages** [reply]
That still depends on the engine you need. I would suggest you check out the minimum system requirements of those applications themselves, if they even list them as they are probably negligible.

**Reply by brainchild0** [reply]
Is slimming down to 4GB disk and 1GB RAM too aggressive?

**Reply by bakerboy448** [reply]
You can probably get by with that.

storage may be an issue with media covers for radarr and sonarr.  You could link those somewhere else if needed; just absolutely do not put the database for those apps on anything but local storage.  SQLLite and network drives means a corrupt database eventually guaranteed.

**Reply by brainchild0** [reply]
Right, administrative data inside; media, including cover art, outside.

**Comment by brainchild0** [comment]
If someone has time, it would be helpful to have a guide for administration of the particular configuration parameters that keep all of the media files in the stores that are external to the VM. These considerations are among the kind that I intended to elucidate in my original inquiry about pitfalls. A contributor or veteran user would be more likely that a newcomer to think through all of the configuration parameters an administrator might need to tweak to accommodate a minimal internal storage footprint, as well as to manage the payload and similar data in an external store. Logs and cache that applications may generate are similar considerations 

For reasons somewhat apparent, I imagine many users would prefer to run in a VM, and so a document consolidating the important issues for such scenarios would be exceedingly valuable.

**Reply by bakerboy448** [reply]
Coming from the Radarr/Sonarr side we see almost no VM use in this manner.  Most use docker or just a VM for their DL client...just for some context.

> If someone has time, it would be helpful to have a guide for administration of the particular configuration parameters that keep all of the media files in the stores that are external to the VM. 

Symlinks may work; otherwise I know users have done docker with different mounts.

All that needs to be moved is `/path/to/app/data/{Arrname}/MediaCovers`  everything else is logs & db for the apps.

For media, the user tells the apps where the library is

**Reply by brainchild0** [reply]
> Coming from the Radarr/Sonarr side we see almost no VM use in this manner. Most use docker or just a VM for their DL client...just for some context.

Would you please clarify the scenario you are describing? Swizzin has no Docker support, so a VM apparently is the only alternative to direct install. I suppose many would prefer to run Swizzin in Docker if it were possible. Are you suggesting installation of Radarr and Sonarr in Docker containers alongside a VM handling Swizzin?



> Symlinks may work.

 Some kind of explanation in the documentation would be useful, even just an entry in a FAQ. If the solution is short and simple, then the explanation also may be short and simple.

**Reply by bakerboy448** [reply]
Was simply speaking generally from the radarr/sonarr side; if the appetite for VMs for the two apps with a combined ~half million users is essentially none; then I can't imagine almost any appetite to use swizzin in a VM; as you know, most have a dedicated mediabox. (edit: or a dedicated mediabox VM...but not a tiny/low power VM)

I'll defer to @flying-sausages , but I'd hedge a bet that customizing partial storage for appdata locations is outside the scope of swizzin docs.

And it is outside the scope for Radarr/Sonarr docs as well...given it's very advanced nature and ease of screwing up/causing problems.

**Reply by brainchild0** [reply]
> Was simply speaking generally from the radarr/sonarr side; if the appetite for VMs for the two apps with a combined ~half million users is essentially none; then I can't imagine almost any appetite to use swizzin in a VM; as you know, most have a dedicated mediabox.

I think  running a bash script as root on a system that has additional uses other than the target of that script is scary, especially if the system runs an  OS image  pre-configured by the vendor of a hardware target, such as an  appliance or SoC. If the only alternative is to run   a small VM, then I imagine doing so might have  appeal to users. If other applications can run in a Docker container, then  it may be preferred, but the question remains of how to run Swizzin, which is the topic of this discussion. A further question is then whether to run any companion applications inside or outside of this VM, and I certainly see value in   a recommendation from someone who understands the mechanics of all these components.

As a very simple starting question, does Swizzin manage Radarr and Sonarr, or do users simply put all the applications on the same system. Is the operation of these applications as companions to Swizzin  affected by the configurations of the various users in the Swizzin deployment?

**Reply by bakerboy448** [reply]
I think there's some confusion here.  Swizzin doesn't run anything.  It's a complication of various scripts as sausage said.

> As a very simple starting question, does Swizzin manage Radarr and Sonarr, or do users simply put all the applications on the same system. 

Again, Swizzin isn't an application at all...it's just a bunch of scripts.  Maybe skim through git real quick so you understand what they do?  Basically they just download and install the app and then auto configure nginx (if installed).  Swizzin doesn't "manage" anything per se other than install/remove.

> Is the operation of these applications as companions to Swizzin affected by the configurations of the various users in the Swizzin deployment?

The apps in question are not multi-user apps and are only used by the "master" user anyway....so the question is moot.

**Reply by brainchild0** [reply]
> I think there's some confusion here. 

Maybe so. I thought   Swizzin provided a lightweight admin frontend, for functions like managing   users, which each has control of a specific torrent set.

**Reply by bakerboy448** [reply]
> lightweight admin frontend, for functions like managing users, which each has control o af specific torrent set.

no front end, everything is around the `box` command. 

but otherwise yes torrent clients can be multi-user

I'll defer to the swizzin docs for more specifics
https://swizzin.ltd/getting-started

**Reply by brainchild0** [reply]
> I think there's some confusion here. Swizzin doesn't run anything.


Fine. I understand this objection as one against some idea that Swizzin is a software component,  separate from various external packages, that persists as a process running on a host. However, I think my comments have meaning separate from such an assumption.

Even understood only as scripts for installation and configuration, such a system plainly makes certain assumptions and manages certain rules about the operation and interaction among the various components. Ultimately, I would understand, a Swizzin deployment is a compilation of loosely coupled pieces, and a more expert understanding of these couplings, applied to the specific question of  running in VM, would be helpful, as I suggested originally, to less familiar users considering this style of deployment.

**Reply by brainchild0** [reply]
> but otherwise yes torrent clients can be multi-user

Where is the multi-user setup documented? Hoes does the admin manage users?

**Reply by bakerboy448** [reply]
> Where is the multi-user setup documented? 

see below;  very, very few apps are multi-user... admittedly it needs to be documented better albiet it'll typically be indicated in the docs within the `Service Management` section, but essentially only the torrent and usenet clients are multiuser... that will be the simplest way to think about it.

> Hoes does the admin manage users?

Did you read the docs? Because I'm not sure you did if you're asking that.
https://swizzin.ltd/getting-started/box-basics

**Reply by brainchild0** [reply]
> Did you read the docs? Because I'm not sure you did if you're asking that.
> https://swizzin.ltd/getting-started/box-basics

You're right. I looked it over briefly a few days ago, but since then I forgot that there is a full manual on the   website outside of the Github repository.



> see below; very, very few apps are multi-user... admittedly it needs to be documented better albiet it'll typically be indicated in the docs within the `Service Management` section, but essentially only the torrent and usenet clients are multiuser... that will be the simplest way to think about it.


Yes, I think a section that explains the features, capabilities, and limitations would be helpful. I think it  is   hard to get a good understanding from the current documentation  of what to expect from the system. For example, what are the use cases?  What can the user do, once the system is running?

Further, given that multi-user support is unavailable for Sonarr, what can a user expect from a secondary user account? Is each user's torrent set integrated into the Sonarr instance, or only the master user's? If Sonarr is normally installed via *box*, what are the consequences for a choice to try to integrate a Docker deployment of Sonarr with a VM deployment of Swizzin, as earlier may have been suggested?

**Reply by bakerboy448** [reply]
Swizzin does not configure Sonarr nor Radarr; that is all up to the user to configure.  

May I suggest reading the swizzin docs and the Radarr/Sonarr Wiki? And again, skimming threw the swizzin scripts

If you want to run docker sonarr along side swizzin installed sonarr then you just need to ensure the ports (sonarr native + docker external), download client categories, and root folders do not conflict.. Beyond that the world is your oyster.  If you don't install sonarr with box then sonarr in docker has literally nothing to do with swizzin,

Honestly, I think you are completely overanalyzing and overthinking this.  Fire up a VM and test things out for yourself...that's the beauty of a VM lol.  In your head it seems fairly clear you're still thinking of swizzin as some kind of management system....it's just a bunch of scripts....for the apps it's just native installs there's nothing fancy about it.

**Reply by brainchild0** [reply]
> May I suggest reading the swizzin docs and the Radarr/Sonarr Wiki? And again, skimming threw the swizzin scripts

Sure, I agree.


> 
> Honestly, I think you are completely overanalyzing and overthinking this. 

Well, maybe I misunderstood some of your earlier comments.

**Reply by flying-sausages** [reply]
> Even understood only as scripts for installation and configuration, such a system plainly makes certain assumptions and manages certain rules about the operation and interaction among the various components. Ultimately, I would understand, a Swizzin deployment is a compilation of loosely coupled pieces, and a more expert understanding of these couplings, applied to the specific question of running in VM, would be helpful, as I suggested originally, to less familiar users considering this style of deployment.

To that, I would say that you should spin up a VM and see what happens ;) But honestly, there's very little difference between a VM with a compatible OS, to a dedicated server with a compatible OS. Feel free to join us in the discord if you'd like to ask about any procedures specifically and we can have a more in-depth look at what you might be concerned about.

**Reply by flying-sausages** [reply]
Also, I wrote these docs about how swizzin itself is structured and what each set of scripts does, this might already give you a good idea of what you can expect. https://swizzin.ltd/dev/structure

**Reply by brainchild0** [reply]
> But honestly, there's very little difference between a VM with a compatible OS, to a dedicated server with a compatible OS.

I agree, but the top-level documentation references the unavailability of a Docker container, but makes no mention of a VM. Yet, it would seem to be a common case, and its absence is conspicuous, at least to me. Although the difference between running in a virtual versus physical machine may seem trivial to you, the end user beginning to explore the project may benefit from some reassurance.

**Comment by kimcosmos** [comment]
actually it says no docker containerS plural. Great move. 
Which is why ppl are spinning up empty lxc debian containers and installing it there. That can then run in proxmox or unraid etc. 
I need a VM for PCI hardware access for gaming. I need swizzin in a container so the hypervisor can cut swizzen down to a minimal machine automatically when my gaming VM pins most of the cores and RAM.
